src/gcp/secretManager.ts

```diff
@@ -63,123 +94,131 @@ export async function secretExists(projectId: string, name: string): Promise<boo
-      origin: api.secretManagerOrigin,
-      data: {
-        payload: {
-          data: Buffer.from(payloadData).toString("base64"),
-        },
+      payload: {
+        data: Buffer.from(payloadData).toString("base64"),
       },
     }
   );
-  const nameTokens = res.body.name.split("/");
   return {
-    secret: {
-      projectId: nameTokens[1],
-      name: nameTokens[3],
-    },
-    versionId: nameTokens[5],
+    ...parseSecretVersionResourceName(res.body.name),
+    state: res.body.state,
   };
 }
 
-export async function grantServiceAgentRole(
+export async function getIamPolicy(secret: Secret): Promise<iam.Policy> {
+  const res = await client.get<iam.Policy>(
+    `projects/${secret.projectId}/secrets/${secret.name}:getIamPolicy`
+  );
+  return res.body;
+}
+
+export async function setIamPolicyBindings(secret: Secret, bindings: iam.Binding[]): Promise<void> {
+  await client.post<{ policy: Partial<iam.Policy> }, iam.Policy>(
+    `projects/${secret.projectId}/secrets/${secret.name}:setIamPolicy`,
+    {
+      policy: {
+        bindings,
+      },
+    },
+    {
+      queryParams: {
+        updateMask: "bindings",
+      },
+    }
+  );
+}
+
+export async function ensureServiceAgentRole(
   secret: Secret,
-  serviceAccountEmail: string,
+  serviceAccountEmails: string[],
   role: string
 ): Promise<void> {
-  const getPolicyRes = await api.request(
-    "GET",
-    `/v1beta1/projects/${secret.projectId}/secrets/${secret.name}:getIamPolicy`,
-    {
-      auth: true,
-      origin: api.secretManagerOrigin,
+  const policy = await module.exports.getIamPolicy(secret);
+  const bindings = policy.bindings || [];
+
+  const newBindings = [];
+  for (const serviceAccountEmail of serviceAccountEmails) {
+    if (
+      !bindings.find(
+        (b: iam.Binding) =>
+          b.role == role &&
+          b.members.find((m: string) => m == `serviceAccount:${serviceAccountEmail}`)
+      )
+    ) {
+      newBindings.push({
+        role: role,
+        members: [`serviceAccount:${serviceAccountEmail}`],
+      });
     }
-  );
+  }
 
-  const bindings = getPolicyRes.body.bindings || [];
-  if (
-    bindings.find(
-      (b: any) =>
-        b.role == role &&
-        b.members.find((m: string) => m == `serviceAccount:${serviceAccountEmail}`)
-    )
-  ) {
-    // binding already exists, short-circuit.
+  if (newBindings.length == 0) {
+    // bindings already exist, short-circuit.
     return;
   }
-  bindings.push({
-    role: role,
-    members: [`serviceAccount:${serviceAccountEmail}`],
-  });
-  await api.request(
-    "POST",
-    `/v1beta1/projects/${secret.projectId}/secrets/${secret.name}:setIamPolicy`,
-    {
-      auth: true,
-      origin: api.secretManagerOrigin,
-      data: {
-        policy: {
-          bindings,
-        },
-        updateMask: {
-          paths: "bindings",
-        },
-      },
-    }
-  );
+
+  bindings.push(...newBindings);
+
+  await module.exports.setIamPolicyBindings(secret, bindings);
```

Comment:
We should also probably return early if the role already exists.

Reply:
I think that's what we do above. Am I mistaking something?

Reply:
This code has changed a bit, but the modified code will short circuit if no changes are necessary.